### ISSUE TITLE
AWS CodeBuild environment_variable conditionally omitted

### ISSUE THREAD

**Issue by hougholi** [terraform-providers/terraform-provider-aws#8936]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.12.1
+ provider.aws v2.14.0
```

### Affected Resource(s)

* aws_codebuild_project

### Terraform Configuration Files

I'm trying to conditionally create an AWS CodeBuild project `environment_variable` if `var.environment` is set to a value. Otherwise the block should be omitted.

```hcl
resource "aws_codebuild_project" "codebuild" {
  name = var.project_name
  service_role  = aws_iam_role.codebuild.arn

  artifacts {
    type = var.s3_artifacts_output ? "S3" : "NO_ARTIFACTS"
    location = var.s3_artifacts_output ? aws_s3_bucket.codebuild.bucket : null
  }

  cache {
    type = "S3"
    location = aws_s3_bucket.codebuild.bucket
  }

  environment {
    compute_type = "BUILD_GENERAL1_SMALL"
    image = "aws/codebuild/standard:2.0"
    type = "LINUX_CONTAINER"
    privileged_mode = true

    environment_variable {
      name = var.environment != null ? "environment" : null
      value = var.environment != null ? var.environment : null
    }
  }

  source {
    type = "S3"
    location = "${aws_s3_bucket.codebuild.bucket}/${var.project_name}.zip"
  }
}
```

### Expected Behavior

If `var.environment` is set to `null`, the `environment_variable` block should be omitted.

### Actual Behavior

If `var.environment` is set to `null`, Terraform tries to create an `environment_variable` block with `type = "PLAINTEXT"`.

```hcl
      + environment {
          + compute_type                = "BUILD_GENERAL1_SMALL"
          + image                       = "aws/codebuild/standard:2.0"
          + image_pull_credentials_type = "CODEBUILD"
          + privileged_mode             = true
          + type                        = "LINUX_CONTAINER"

          + environment_variable {
              + type = "PLAINTEXT"
            }
        }
```
```
Error: [ERROR] Error updating CodeBuild project (arn:aws:codebuild:us-east-1:111111111111:project/codebuild): InvalidParameter: 2 validation error(s) found.
- missing required field, UpdateProjectInput.Environment.EnvironmentVariables[0].Name.
- missing required field, UpdateProjectInput.Environment.EnvironmentVariables[0].Value.
```

### Steps to Reproduce

1. Set `var.environment` to `null`
2. `terraform apply`

### Important Factoids

Even if `type` is set to `null` in the `environment_variable` block, Terraform still tries to create an `environment_variable` block with `type = "PLAINTEXT"`

```hcl
resource "aws_codebuild_project" "codebuild" {
  name = var.project_name
  service_role  = aws_iam_role.codebuild.arn

  artifacts {
    type = var.s3_artifacts_output ? "S3" : "NO_ARTIFACTS"
    location = var.s3_artifacts_output ? aws_s3_bucket.codebuild.bucket : null
  }

  cache {
    type = "S3"
    location = aws_s3_bucket.codebuild.bucket
  }

  environment {
    compute_type = "BUILD_GENERAL1_SMALL"
    image = "aws/codebuild/standard:2.0"
    type = "LINUX_CONTAINER"
    privileged_mode = true

    environment_variable {
      name = var.environment != null ? "environment" : null
      value = var.environment != null ? var.environment : null
      type = null
    }
  }

  source {
    type = "S3"
    location = "${aws_s3_bucket.codebuild.bucket}/${var.project_name}.zip"
  }
}
```
**Output**
```hcl
      + environment {
          + compute_type                = "BUILD_GENERAL1_SMALL"
          + image                       = "aws/codebuild/standard:2.0"
          + image_pull_credentials_type = "CODEBUILD"
          + privileged_mode             = true
          + type                        = "LINUX_CONTAINER"

          + environment_variable {
              + type = "PLAINTEXT"
            }
        }
```

**Comment by bflad** [comment]
Hi @hougholi  👋 This issue is resolved in Terraform 0.12, which supports new functionality in the configuration language aimed at solving problems like these. The new [`dynamic` block syntax](https://www.terraform.io/docs/configuration/expressions.html#dynamic-blocks) can be used to dynamically generate configuration blocks and their arguments. These can be combined with the new [`null` value](https://www.terraform.io/docs/configuration/expressions.html#null), which can be used to omit arguments as if they were not defined in the configuration at all.

You can accomplish something like you're looking for with the following example:

```hcl
terraform {
  required_providers {
    aws = "2.21.1"
  }
  required_version = "0.12.5"
}

provider "aws" {
  region = "us-east-2"
}

resource "aws_iam_role" "test" {
  assume_role_policy = jsonencode({
    Version = "2012-10-17"
    Statement = [{
      Action = "sts:AssumeRole"
      Effect = "Allow",
      Principal = {
        Service = "codebuild.amazonaws.com"
      }
    }]
  })
}

resource "aws_iam_role_policy" "test" {
  role = aws_iam_role.test.name

  policy = jsonencode({
    Version = "2012-10-17",
    Statement = [
      {
        Action = [
          "logs:CreateLogGroup",
          "logs:CreateLogStream",
          "logs:PutLogEvents",
        ]
        Effect   = "Allow",
        Resource = ["*"]
      },
      {
        Action = [
          "ec2:CreateNetworkInterface",
          "ec2:CreateNetworkInterfacePermission",
          "ec2:DescribeDhcpOptions",
          "ec2:DescribeNetworkInterfaces",
          "ec2:DeleteNetworkInterface",
          "ec2:DescribeSubnets",
          "ec2:DescribeSecurityGroups",
          "ec2:DescribeVpcs",
        ]
        Effect   = "Allow"
        Resource = "*"
      },
      {
        Action   = ["s3:*"],
        Effect   = "Allow",
        Resource = "*"
      },
    ]
  })
}

variable "environment_variables" {
  default = {}
  type    = map(string)
}

resource "aws_codebuild_project" "test" {
  name         = "test"
  service_role = "${aws_iam_role.test.arn}"

  artifacts {
    type = "NO_ARTIFACTS"
  }

  environment {
    compute_type = "BUILD_GENERAL1_SMALL"
    image        = "2"
    type         = "LINUX_CONTAINER"

    dynamic "environment_variable" {
      for_each = var.environment_variables

      content {
        name  = environment_variable.key
        value = environment_variable.value
      }
    }
  }

  source {
    location = "https://github.com/hashicorp/packer.git"
    type     = "GITHUB"
  }
}
```

When left with the default of an empty map, it omits the configuration blocks:

```console
$ terraform state show aws_codebuild_project.test
# aws_codebuild_project.test:
resource "aws_codebuild_project" "test" {
# ... omitted for brevity ...
    environment {
        compute_type                = "BUILD_GENERAL1_SMALL"
        image                       = "2"
        image_pull_credentials_type = "CODEBUILD"
        privileged_mode             = false
        type                        = "LINUX_CONTAINER"
    }
# ... omitted for brevity ....
}
```

As a contrived example, adjusting the variable as such:

```hcl
variable "environment_variables" {
  default = {
    key1 = "value1"
    key2 = "value2"
  }
  type = map(string)
}
```

Shows Terraform will automatically create two configuration blocks on apply:

```console
$ terraform apply
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aws_codebuild_project.test will be updated in-place
  ~ resource "aws_codebuild_project" "test" {
# ... omitted for brevity ...
      + environment {
          + compute_type                = "BUILD_GENERAL1_SMALL"
          + image                       = "2"
          + image_pull_credentials_type = "CODEBUILD"
          + privileged_mode             = false
          + type                        = "LINUX_CONTAINER"

          + environment_variable {
              + name  = "key1"
              + type  = "PLAINTEXT"
              + value = "value1"
            }
          + environment_variable {
              + name  = "key2"
              + type  = "PLAINTEXT"
              + value = "value2"
            }
        }
      - environment {
          - compute_type                = "BUILD_GENERAL1_SMALL" -> null
          - image                       = "2" -> null
          - image_pull_credentials_type = "CODEBUILD" -> null
          - privileged_mode             = false -> null
          - type                        = "LINUX_CONTAINER" -> null
        }
# ... omitted for brevity ...
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Enjoy! 🚀

**Comment by hougholi** [comment]
Thanks @bflad!
